client/OPUNetGameSelectWnd.cpp

```diff
@@ -357,20 +357,7 @@ void OPUNetGameSelectWnd::OnTimer()
 
 	if ((externalPort == 0) && (numEchoRequestsSent < MaxEchoAttempt))
 	{
-		echoTick++;
-		if (echoTick >= EchoTickInterval)
-		{
-			if (numEchoRequestsSent == 0)
-			{
-				internalPort = opuNetTransportLayer->GetPort();
-			}
-
-			echoTick = 0;
-			numEchoRequestsSent++;
-
-			// Request external address
-			opuNetTransportLayer->GetExternalAddress();
-		}
+		UpdateEchoRequest();
```

Comment:
I think I might name this one `RequestExternalAddress`, or similar. The name "echo" is more about how it get's the external address, rather than what it's doing. A computer has no way by itself to know what the external WAN IP of the router is. It gets this by sending a packet through the router, having addresses translated by NAT, and then receiving an echo back with the external address packaged into the message (where it is safe from further NAT translation). That can be used to determine if NAT is in effect, and gives some indication as to what security filtering characteristics the NAT translation has, particularly when getting multiple replies back.

As the characteristic is of the local router, this information is not expected to change once determined. Not many people upgrade their router while playing a network game. Hence it makes a few attempts early on to detect NAT type, and then stops.